de.dlr.sc.virsat.model.extension.requirements.ui/src/de/dlr/sc/virsat/model/extension/requirements/ui/snippet/UiSnippetCustomRequirementsAttributeTable.java

```diff
@@ -0,0 +1,369 @@
+				if (reqObject.getType().getFullQualifiedName()
+						.equals(Requirement.FULL_QUALIFIED_CATEGORY_NAME)) {
+					RequirementType requirementType = new Requirement(reqObject).getReqType();
+					if (requirementType != null) {
+						requirementTypes.add(requirementType);
+						int numberAttributes = requirementType.getAttributes().size();
+						maxNumberAttributes = Math.max(maxNumberAttributes, numberAttributes);
+					}
+				}
+			}
+
+			// Add necessary table columns
+			for (int i = 0; i < maxNumberAttributes; i++) {
+				String columnName = "";
+				for (RequirementType requirementType : requirementTypes) {
+
+					if (requirementType.getAttributes().size() > i) {
+						String name = requirementType.getAttributes().get(i).getName();
+						boolean notYetSpecified = !columnName.contains(name);
+						
+						// Add separator if column is used for different attributes
+						if (!columnName.equals("") && notYetSpecified) {
+							columnName += COLUMN_ATTRIBUTE_SEPARATOR;
+						}
+
+						if (notYetSpecified) {
+							columnName += name;
+						}
+					}
+				}
+
+				if (attColumns.size() > i) {
+					attColumns.get(i).getColumn().setText(columnName);
+				} else {
+					TableViewerColumn newColumn = (TableViewerColumn) createDefaultColumn(columnName);
+					newColumn.getColumn().addControlListener(this);
+					newColumn.setEditingSupport(new RequirementsAttributeValuePerColumnEditingSupport(editingDomain, columnViewer, i));
+					attColumns.add(newColumn);
+				}
+			}
+		}
+		restoreColumnWitdh();
+	}
+	
+
+	/**
+	 * this method get the label provider
+	 * 
+	 * @return the table column provider
+	 */
+	protected ITableLabelProvider getTableLabelProvider() {
+		VirSatTransactionalAdapterFactoryLabelProvider labelProvider;
+		labelProvider = new RequirementsAttributeLabelProvider(adapterFactory);
+		return labelProvider;
+	}
+
+	/**
+	 * Refresh the table if a requirement with new type was added
+	 * 
+	 * @param editingDomain the editing domain
+	 */
+	protected void refreshTable(EditingDomain editingDomain) {
+		createTableColumns(editingDomain);
+		ITableLabelProvider labelProvider = getTableLabelProvider();
+		columnViewer.setLabelProvider(labelProvider);
+	}
+	
+	
+	/**
+	 * this method creates the add button and his functionality
+	 * 
+	 * @param toolkit the toolkit to be used to create the button
+	 * @param editingDomain the editing domain an which the button might act on
+	 * @param compositeButtons the composite in which the button should be placed
+	 */
+	protected void createAddButton(FormToolkit toolkit, EditingDomain editingDomain, Composite compositeButtons) {
+		if ((style & STYLE_ADD_BUTTON) != 0) {
+			buttonAdd = toolkit.createButton(compositeButtons, BUTTON_ADD_TEXT + getTypeInformation(), SWT.PUSH);
+			buttonAdd.addSelectionListener(new SelectionListener() {
+				@Override
+				public void widgetSelected(SelectionEvent e) {
+
+					Concept activeConcept = acHelper.getConcept(conceptId);

```

Comment:
The method is a bit long.
Could we maybe extract something like a `createDialog(...)` method encapsulating the logic for creating and setting up the dialog?